### ISSUE TITLE
Sporadic test failures with sqlite and python3

### ISSUE THREAD

**Issue by drnlm** [sqlobject/sqlobject#135]
First observed when trying to build sqlobject 3.3.0 for debian, but also present on current master

The sqlobject test suite sporadically fails with python 3.5 and python 3.6 in test_unicode.py.

The python 3.5 failure is less common, only occurring about once every 10 test runs on my machine, while the python 3.6 failure happens every 3 or 4 runs. I haven't been able to trigger the failure with python 2.

Python 3.6 fails with
```
__________________________ ERROR at setup of test_create __________________________

self = <sqlobject.sqlite.sqliteconnection.SQLiteConnection object at 0xf6844bcc>
conn = <sqlite3.Connection object at 0xf6467d20>
cursor = <sqlite3.Cursor object at 0xf59870e0>
query = "INSERT INTO so_test_unicode (so_count, col) VALUES (0, '\xf0')"

    def _executeRetry(self, conn, cursor, query):
        if self.debug:
            self.printDebug(conn, query, 'QueryR')
        try:
>           return cursor.execute(query)
E           sqlite3.OperationalError: no such table: so_test_unicode

sqlite/sqliteconnection.py:220: OperationalError

During handling of the above exception, another exception occurred:

    def setup():
        global items
        items = []
        setupClass(SOTestUnicode)
        for i, s in enumerate(data):
>           items.append(SOTestUnicode(so_count=i, col=s))

tests/test_unicode.py:25: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
main.py:1301: in __init__
    self._create(id, **kw)
main.py:1352: in _create
    self._SO_finishCreate(id)
main.py:1378: in _SO_finishCreate
    id, names, values)
dbconnection.py:488: in queryInsertID
    names, values)
dbconnection.py:342: in _runWithConnection
    val = meth(conn, *args)
sqlite/sqliteconnection.py:257: in _queryInsertID
    self._executeRetry(conn, c, q)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sqlobject.sqlite.sqliteconnection.SQLiteConnection object at 0xf6844bcc>
conn = <sqlite3.Connection object at 0xf6467d20>
cursor = <sqlite3.Cursor object at 0xf59870e0>
query = "INSERT INTO so_test_unicode (so_count, col) VALUES (0, '\xf0')"

    def _executeRetry(self, conn, cursor, query):
        if self.debug:
            self.printDebug(conn, query, 'QueryR')
        try:
            return cursor.execute(query)
        except self.module.OperationalError as e:
>           raise dberrors.OperationalError(ErrorMessage(e))
E           sqlobject.dberrors.OperationalError: no such table: so_test_unicode

sqlite/sqliteconnection.py:222: OperationalError
```

Python 3.5 fails with
```
__________________________ ERROR at setup of test_create __________________________

    def setup():
        global items
        items = []
        setupClass(SOTestUnicode)
        for i, s in enumerate(data):
>           items.append(SOTestUnicode(so_count=i, col=s))

tests/test_unicode.py:25: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
main.py:1301: in __init__
    self._create(id, **kw)
main.py:1352: in _create
    self._SO_finishCreate(id)
main.py:1381: in _SO_finishCreate
    self._init(id)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <SOTestUnicode 2 so_count=1 col='test'>, id = 2, connection = None
selectResults = None

    def _init(self, id, connection=None, selectResults=None):
        assert id is not None
        # This function gets called only when the object is
        # created, unlike __init__ which would be called
        # anytime the object was returned from cache.
        self.id = id
        self._SO_writeLock = threading.Lock()
    
        # If no connection was given, we'll inherit the class
        # instance variable which should have a _connection
        # attribute.
        if (connection is not None) and \
                (getattr(self, '_connection', None) is not connection):
            self._connection = connection
            # Sometimes we need to know if this instance is
            # global or tied to a particular connection.
            # This flag tells us that:
            self.sqlmeta._perConnection = True
    
        if not selectResults:
            dbNames = [col.dbName for col in self.sqlmeta.columnList]
            selectResults = self._connection._SO_selectOne(self, dbNames)
            if not selectResults:
                raise SQLObjectNotFound(
                    "The object %s by the ID %s does not exist" % (
>                       self.__class__.__name__, self.id))
E               sqlobject.main.SQLObjectNotFound: The object SOTestUnicode by the ID 2 does not exist

main.py:985: SQLObjectNotFound
```

**Comment by phdru** [comment]
I regularly see the problem at Travis, but only at Travis. I could never reproduce it at home or at AppVeyor.

[Works](https://travis-ci.org/sqlobject/sqlobject/builds/244049853)… [Fails](https://travis-ci.org/sqlobject/sqlobject/jobs/248152002)… and so on.

Not a slightest idea what's up. :-(

**Comment by drnlm** [comment]
After noticing that the debian package builds fail very reliably on python 3.6, I've determined that, if the dbs_data.tmp created by InstalledTestDatabase exists from a previous test, the python 3.6 tests behave as described earlier, failing every 3 or 4 runs on my machine. If the file is deleted before running the tests, the python 3.6 tests fail a lot more consistently  - not always in exactly the same place, but it almost always fails. 

Playing around with travis, I see the same behaviour by forcing the test suite to run twice - https://travis-ci.org/drnlm/sqlobject/jobs/255288269, where the first fails but the second (with dbs_data.tmp present) succeeds (I've manually rerun that particular job several times to confirm the behaviour).

This seems to have no impact on whether the python 3.5 tests fail or succeed, though.

**Comment by phdru** [comment]
Very interesting, thank you for finding! I'll try to experiment at home.

**Comment by phdru** [comment]
Well, I cannot reproduce that at home. The only difference I see is that I run 32-bit Debian 8.8 and I've compiled Python 3.6 myself.

**Comment by phdru** [comment]
Reproduced at home. Removing dbs_data.tmp before running tests reproduces the problem (python 3.6 but not other) with some probability:

```
[sqlite-memory]
commands =
    {[testenv]commands}
    rm -f dbs_data.tmp
    pytest -D sqlite:/:memory:?debug=1
```

**Comment by phdru** [comment]
During experiments with sqlite-memory tests I changed the order of tests — similar to what I've done with Postgres tests — that fixed problems with py36-sqlite-memory and created a problem with [py27-sqlite-memory](https://travis-ci.org/sqlobject/sqlobject/jobs/256473839). The problem seems to be 100% reproducible. Not sure if it is the same problem, or similar, or completely different.

**Comment by drnlm** [comment]
I think the problem may be due to a reference created by test_transaction_commit_sync somehow.  
Playing with various gc settings, I started getting a
```Exception ignored in: <bound method Transaction.__del__ of <sqlobject.dbconnection.Transaction object at 0xf59d84cc>>
Traceback (most recent call last):
  File "/tmp/sqlobject/sqlobject/dbconnection.py", line 925, in __del__
    self.rollback()
  File "/tmp/sqlobject/sqlobject/dbconnection.py", line 872, in rollback
    self._connection.rollback()
sqlite3.ProgrammingError: Cannot operate on a closed database.
```

which I was able to trace back to the transaction created in that test.

Adding an explicit trans.commit(close=True) to that test avoids both that warning and makes the tests pass reliably for me with python 2.7, 3.5 and 3.6 (or at least very hard to trigger - I have now run the tests more than 20 times with each version without a failure). 

I suspect that this is a result 'test_transaction_delete_with_close' being skipped with the sqlite memory tests.

**Comment by phdru** [comment]
I've already traced the problem in `__del__` to [atexit function](https://github.com/sqlobject/sqlobject/blob/master/sqlobject/dbconnection.py#L105) but not deeper.

Interesting finding. I'll add `trans.commit(close=True)` to the tests.

**Comment by phdru** [comment]
Instead of committing I decided to try to [repeat](https://github.com/sqlobject/sqlobject/commit/48ddb3a762f62b03843db11a9d3fc878ef2038fd#diff-7361aee7fb3e4b0a01f06c2ae62c2d5eL63) code from the [skipped test](https://github.com/sqlobject/sqlobject/blob/master/sqlobject/tests/test_transactions.py#L88). Seems to be [working now](https://travis-ci.org/sqlobject/sqlobject/builds/259977695)!

**Comment by phdru** [comment]
Fixed by 48ddb3a. Many thanks to @drnlm!
